bootloader/ubootenv/env.go

```diff
@@ -20,120 +20,94 @@
-	buf := bytes.NewBuffer(nil)
-	binary.Write(buf, binary.LittleEndian, &u)
-	return buf.Bytes()
-}
-
-// Create a new empty uboot env file with the given size
-func Create(fname string, size int) (*Env, error) {
-	f, err := os.Create(fname)
-	if err != nil {
-		return nil, err
-	}
-	defer f.Close()
-
-	env := &Env{
-		fname: fname,
-		size:  size,
-		data:  make(map[string]string),
-	}
-
-	return env, nil
+	// mostly for tests or debugging
+	String() string
+	Size() int
 }
 
 // OpenFlags instructs open how to alter its behavior.
 type OpenFlags int
 
+// TODO: support "binary" format too?
+
+// Format is the uboot environment format
+type Format int
+
 const (
-	// OpenBestEffort instructs OpenWithFlags to skip malformed data without returning an error.
+	// OpenBestEffort instructs OpenWithFlags to skip malformed data without
+	// returning an error.
 	OpenBestEffort OpenFlags = 1 << iota
-)
 
-// Open opens a existing uboot env file
-func Open(fname string) (*Env, error) {
-	return OpenWithFlags(fname, OpenFlags(0))
-}
-
-// OpenWithFlags opens a existing uboot env file, passing additional flags.
-func OpenWithFlags(fname string, flags OpenFlags) (*Env, error) {
-	f, err := os.Open(fname)
-	if err != nil {
-		return nil, err
-	}
-	defer f.Close()
+	// OpenIgnoreComments instructs OpenWithFlags to read lines/entries that
+	// start with a "#" character, the same way mkenvimage does with input files
+	OpenIgnoreComments
 
-	contentWithHeader, err := ioutil.ReadAll(f)
-	if err != nil {
-		return nil, err
-	}
-	crc := readUint32(contentWithHeader)
+	// NativeFormat is the "uboot native" format which has a CRC, delimits vars
+	// with '\0' and has padding on it
+	NativeFormat Format = iota
+	// TextFormat is the simple text format for vars
+	TextFormat
+)
 
-	payload := contentWithHeader[headerSize:]
-	actualCRC := crc32.ChecksumIEEE(payload)
-	if crc != actualCRC {
-		return nil, fmt.Errorf("cannot open %q: bad CRC %v != %v", fname, crc, actualCRC)
+// Create will initialize a uboot env file with the given options. For
+// TextFormat the size parameter is ignored.
+func Create(fname string, format Format, size int) (Env, error) {
+	switch format {
+	case NativeFormat:
+		return createNative(fname, size)
+	case TextFormat:
+		return createText(fname)
 	}
 
-	if eof := bytes.Index(payload, []byte{0, 0}); eof >= 0 {
-		payload = payload[:eof]
-	}
+	return nil, fmt.Errorf("unsupported format %v", format)
+}
 
-	data, err := parseData(payload, flags)
-	if err != nil {
-		return nil, err
-	}
+// Open opens a existing uboot env file
+func Open(fname string, fmt Format) (Env, error) {
+	return OpenWithFlags(fname, fmt, OpenFlags(0))
+}
 
-	env := &Env{
-		fname: fname,
-		size:  len(contentWithHeader),
-		data:  data,
+// OpenWithFlags opens a existing uboot env file, passing additional flags.
+func OpenWithFlags(fname string, format Format, flags OpenFlags) (Env, error) {
+	switch format {
+	case NativeFormat:
+		return openNativeFormat(fname, flags)
```

Comment:
where is this defined?

Reply:
d'oh I missed the files when committing since they're new files. That probably serves me right for trying to write code so early...